### PR TITLE
Add variadic versions for catch

### PR DESCRIPTION
**Pull request by juanpaucar**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://bluebirdjs.com/docs/api/catch.html#filtered-catch
- [ ] Increase the version number in the header if appropriate. <-- **NOT SURE** 
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Hi, so I wanted to extend the typings definition for `bluebird` since as noted in the [link](http://bluebirdjs.com/docs/api/catch.html#filtered-catch) it can be variadic and accept several types of errors.

image

Currently, left variadic functions [are not supported properly](https://github.com/Microsoft/TypeScript/issues/5453). So, a workaround is to implement for several functions definitions. However, I'm not sure if my approach is the correct because of the reasons below:

thanks @lhecker for your help. I just learned a lot more thanks to you 👍

### CODE REVIEW

**Review by lhecker — CHANGES_REQUESTED** [review]
@juanpaucar I took the liberty to fix your code. 🙂 You can see the diff [here](https://gist.github.com/lhecker/2e7ae57ebbc595639ca52b2c7824e93a).
If you're okay with those changes you could apply them using e.g.:

```sh
curl "https://gist.githubusercontent.com/lhecker/2e7ae57ebbc595639ca52b2c7824e93a/raw/f28e855605e0d801f5b7d16f8cb3358b0a6d81c7/bluebird.dt.diff" | git apply
```

I'd suggest `git commit --amend`-ing it so we have a clear functioning starting point for this PR (but if you do don't forget to [remove the dangling commit](https://stackoverflow.com/a/4528593) afterwards).

<hr>

Anyways… To make it simple, the difference between overwriting the `.prototype` of a function ("old syntax") and using the `class A extends B` syntax is that the latter also properly sets the `constructor` and `__proto__` properties. But apart from that there are only syntactic differences.

Since TypeScript is ES6 based and uses the class syntax for type checking the loosely typed overwriting of `.prototype` cannot automatically be supported for bluebird.

<hr>

```js
fooProm = fooProm.catch(new CustomError1(), new CustomError2(), error => {});
```

As far as I can see it bluebird does not support passing error _instances_ as a filter to `.catch()`. (This is also the reason why I removed it from my .diff above).

<hr>

```js
fooProm = fooProm.catch(CustomError1, CustomError2, error => {});
```

Your variadic catch methods were basically: `catch<E1 extends Error, ...>(e1: E1, ...)`. This specifies a method signature which takes _instances_ of type `E1`. What we want instead are class references which when instantiated return errors of type `E1`.
The existing `catch()` methods already solved this by taking a `ErrorClass: new (...args: any[]): E`. In my .diff above I solved it similarly.

**Comment by juanpaucar** [comment]
@lhecker Thanks I just figured that too. `e1: new (...args: any[]) => E1` was missing. However I think that changing it to filter is more appropriate. I also notice that you updated more methods to make them variadic. Should I include them too?

**Comment by lhecker** [comment]
> Should I include them too?

Yes definitely! If we fix something we should fix all of it, right? 🙂
BTW: My .diff also includes some simplifications by changing the general filter signature to

```
filter1: (new (...args: any[]) => E1) | ((error: any) => boolean) | object
```

instead of having 3 seperate methods for that. This also better matches the bluebird API docs.

**Comment by juanpaucar** [comment]
Thanks. I just updated this PR description's

**Comment by juanpaucar** [comment]
@lhecker I have another question and I think I just committed a silly mistake. Should i make this PR here or in [this repo](https://github.com/types/npm-bluebird)

**Comment by lhecker** [comment]
@juanpaucar If I remember correctly the [github.com/types](https://github.com/types) project was the original provider for the `@types/...` packages on npm. But this changed a year back or so, when DefinitelyTyped took over those packages as the provider due to popularity.
Back then this repo was missing bluebird 3.0 typings so I basically "forked" `npm-bluebird` and contributed it to DefinitelyTyped.
You probably did the right thing contributing to this project since `npm-bluebird` would only be interesting to contribute to if you want to fix projects with old legacy dependencies to it.

File: types/bluebird/index.d.ts [hunk 1]
```diff
@@ -66,23 +66,130 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
...
-  catch<E extends Error, U>(ErrorClass: new (...args: any[]) => E, onReject: (error: E) => U | Bluebird.Thenable<U>): Bluebird<U | R>;
-  caught<E extends Error, U>(ErrorClass: new (...args: any[]) => E, onReject: (error: E) => U | Bluebird.Thenable<U>): Bluebird<U | R>;
-
-  catch(predicate: Object, onReject: (error: any) => R | Bluebird.Thenable<R> | void | Bluebird.Thenable<void>): Bluebird<R>;
-  caught(predicate: Object, onReject: (error: any) => R | Bluebird.Thenable<R> | void | Bluebird.Thenable<void>): Bluebird<R>;
-
-  catch<U>(predicate: Object, onReject: (error: any) => U | Bluebird.Thenable<U>): Bluebird<U | R>;
-  caught<U>(predicate: Object, onReject: (error: any) => U | Bluebird.Thenable<U>): Bluebird<U | R>;
+  catch<E1 extends Error>(
+    filter1: (new (...args: any[]) => E1) | ((error: any) => boolean) | Object,
```
Comment (lhecker) [inline types/bluebird/index.d.ts]: I believe you said you're new to TypeScript? If so: Did you know that there's quite a difference between `Object` and `object`? The former acts like `any` which is a behavior we don't really want, right? The latter on the other hand only allows actual objects to be passed.
The problem is that by using `object` we'd increase the minimal TypeScript requirement to version 2.2 and I'm not sure if the owners of this repo would want this…

**Review by lhecker — APPROVED** [review]

File: types/bluebird/index.d.ts (hunk 1 above)
Comment (juanpaucar) [inline types/bluebird/index.d.ts]: Yes, I knew it. I actually had a comment with that observation and explaining why I passed `Object` instead of `object` we would still be missing some cases for non primitives here but I think that `Object` may work until the minimum version be 2.2. Do you think that it's worth to keep `Object` or should be better if removed?

Comment (juanpaucar) [inline types/bluebird/index.d.ts]: ping @lhecker

Comment (lhecker) [inline types/bluebird/index.d.ts]: Oh sorry @juanpaucar. We have to keep `Object` for now, since otherwise we break compatibility with the previous bluebird.d.ts version. It _might_ be worth it to amend a `// TODO disallow non-objects` comment to this PR, just like #16572 did so we don't forget it.

Comment (juanpaucar) [inline types/bluebird/index.d.ts]: No problem. I just added some TODOs where the `Object` was added

**Comment by juanpaucar** [comment]
@lhecker Is there any other change that prevents this PR from being merged?

**Comment by lhecker** [comment]
@juanpaucar Not from my side. I believe it's just because Microsoft et. al. are a bit slow in merging the PRs currently (they've been quite fast with it in the past). I sadly don't know why that's the case.

Maybe @DanielRosenwasser could help us again? 🙂

File: types/bluebird/index.d.ts [hunk 2]
```diff
@@ -116,11 +225,40 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
 
   /**
    * Like `.catch()` but rethrows the error
-   */
-  tapCatch<U>(onReject: (error?: any) => U | Bluebird.Thenable<U> | void | Bluebird.Thenable<void>): Bluebird<R>;
-  tapCatch<U, E extends Error>(ErrorClass: new (...args: any[]) => E, onReject: (error: E) => U | Bluebird.Thenable<U> | void | Bluebird.Thenable<void>): Bluebird<R>;
-  tapCatch<U>(predicate: (error?: any) => boolean, onReject: (error?: any) => U | Bluebird.Thenable<U> | void | Bluebird.Thenable<void>): Bluebird<R>;
-
+   * TODO: disallow non-objects
```
Comment (DanielRosenwasser) [inline types/bluebird/index.d.ts]: What do you mean? If you just switch from `Object` to `object` and add the 

```ts
// TypeScript Version: 2.2.
```

header, it should just work.

File: types/bluebird/index.d.ts [hunk 3]
```diff
@@ -116,11 +225,40 @@ declare class Bluebird<R> implements Bluebird.Thenable<R>, Bluebird.Inspection<R
...
+    onReject: (error: E1 | E2 | E3) => U | Bluebird.Thenable<U>,
+  ): Bluebird<R>;
+  tapCatch<U, E1 extends Error, E2 extends Error, E3 extends Error, E4 extends Error>(
+    filter1: (new (...args: any[]) => E1) | ((error: any) => boolean) | Object,
+    filter2: (new (...args: any[]) => E2) | ((error: any) => boolean) | Object,
+    filter3: (new (...args: any[]) => E3) | ((error: any) => boolean) | Object,
+    filter4: (new (...args: any[]) => E4) | ((error: any) => boolean) | Object,
+    onReject: (error: E1 | E2 | E3 | E4) => U | Bluebird.Thenable<U>,
+  ): Bluebird<R>;
+  tapCatch<U, E1 extends Error, E2 extends Error, E3 extends Error, E4 extends Error, E5 extends Error>(
```
Comment (DanielRosenwasser) [inline types/bluebird/index.d.ts]: The more specific overloads should come first. That means the overloads that take *more arguments* (i.e. this one) should come first, and the overload that takes the fewest arguments should come last (i.e. the overload that takes one argument should come last.

TL;DR: These need to be reversed. Likewise for `caught` and `catch`.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: types/bluebird/index.d.ts (hunk 2 above)
Comment (juanpaucar) [inline types/bluebird/index.d.ts]: I know. But i thought that we'd like to still support version 2.0 as well. Should i make it compatible only for 2.2 and newer (changing `Object` for `object`), is that ok?

\cc @lhecker

File: types/bluebird/index.d.ts (hunk 3 above)
Comment (juanpaucar) [inline types/bluebird/index.d.ts]: Done :)

File: types/bluebird/index.d.ts (hunk 2 above)
Comment (juanpaucar) [inline types/bluebird/index.d.ts]: This is the only doubt I have. Drop support for 2.0 and use `object` or leave it for future changes to drop support and make the change
\cc @lhecker @DanielRosenwasser

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by juanpaucar** [comment]
@jcready I'm sorry. I totally missed that part. Do you want to make the change and open a PR? or should I do it?

**Comment by jcready** [comment]
@juanpaucar honestly, it's kinda impossible to correctly refine the types that get passed to `onReject`. Like I said, the *only* way to safely refine the argument that gets passed to `onReject` is if the only thing you're accepting for the filters are classes which extend `Error`. But since Bluebird allows you to pass object or function predicates as well, you're essentially left with being forced to use `any` for the type which gets passed to `onReject` which is incredibly unfortunate. I don't see a way to refine the argument passed to `onReject` without it being incorrect some of the time. :hurtrealbad:

**Comment by juanpaucar** [comment]
@jcready my friend was taking a look at this and openned a PR #17306 . Can you take a look?
